pkg/scheduler/objects/preemption.go

```diff
@@ -0,0 +1,875 @@
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package objects
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/apache/yunikorn-core/pkg/common/resources"
+	"github.com/apache/yunikorn-core/pkg/log"
+	"github.com/apache/yunikorn-core/pkg/plugins"
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/api"
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
+)
+
+var (
+	preemptAttemptFrequency        = 15 * time.Second
+	preemptCheckConcurrency        = 10
+	scoreFitMax             uint64 = 1 << 32
+	scoreOriginator         uint64 = 1 << 33
+	scoreNoPreempt          uint64 = 1 << 34
+	scoreUnfit              uint64 = 1 << 35
+)
+
+// Preemptor encapsulates the functionality required for preemption victim selection
+type Preemptor struct {
+	application     *Application        // application containing ask
+	queue           *Queue              // queue to preempt for
+	queuePath       string              // path of queue to preempt for
+	headRoom        *resources.Resource // current queue headroom
+	preemptionDelay time.Duration       // preemption delay
+	ask             *AllocationAsk      // ask to be preempted for
+	iterator        NodeIterator        // iterator to enumerate all nodes
+	nodesTried      bool                // flag indicating that scheduling has already been tried on all nodes
+
+	// lazily-populated work structures
+	allocationsByQueue map[string]*QueuePreemptionSnapshot // map of queue snapshots by queue path
+	queueByAlloc       map[string]*QueuePreemptionSnapshot // map of queue snapshots by allocationID
+	allocationsByNode  map[string][]*Allocation            // map of allocation by nodeID
+	nodeAvailableMap   map[string]*resources.Resource      // map of available resources by nodeID
+}
+
+// QueuePreemptionSnapshot is used to track a snapshot of a queue for preemption
+type QueuePreemptionSnapshot struct {
+	Parent             *QueuePreemptionSnapshot // snapshot of parent queue
+	QueuePath          string                   // fully qualified path to queue
+	AllocatedResource  *resources.Resource      // allocated resources
+	PreemptingResource *resources.Resource      // resources currently flagged for preemption
+	MaxResource        *resources.Resource      // maximum resources for this queue
+	GuaranteedResource *resources.Resource      // guaranteed resources for this queue
+	PotentialVictims   []*Allocation            // list of allocations which could be preempted
+}
+
+// NewPreemptor creates a new preemptor. The preemptor itself is not thread safe, and assumes the application lock is held.
```

Comment:
Nit: we already have `RequiredNodePreemptor`, could we call this `GenericPreemptor` to emphasize the difference?

Reply:
I'm hesitant to do this. In fact, there was a point during the development cycle where I considered folding the required node preemptor back into this one (making it the singular preemptor). I'm not generally a fan of longer names when shorter ones will do. The required node preemptor is of course the odd one out now, so the longer name makes sense.